text/2018-01-07-deque-proof.md

```diff
@@ -0,0 +1,953 @@
+
+A Chase-Lev deque consists of (1) the `bottom` index for the worker, (2) the `top` index for both
+the worker and the stealers, and (3) the underlying `buffer` that contains items. When the buffer is
+too small or too large, the worker may replace it with one with more appropriate size:
+
+```rust
+struct<T: Send> Inner<T> {
+    bottom: AtomicUsize,
+    top: AtomicUsize,
+    buffer: Ptr<Buffer<T>>,
+}
+
+impl<T: Send> Inner<T> {
+    ...
+}
+```
+
+The `Deque` and `Stealer` structs have a counted reference to the inner implementation, and delegate
+method invocations to it:
+
+```rust
+struct<T: Send> Deque<T> {
+    inner: Arc<Inner<T>>,
+}
+
+impl<T:Send> Deque<T> {
+    fn push(...) {
+        self.inner.push(...)
+    }
+
+    ...
+}
+
+...
+```
+
+The pseudocode for `Inner<T>`'s `fn push()`, `fn pop()`, and `fn steal()` is as follows:
+
+```rust
+pub fn push(&self, t: T) {
+    'L101: let b = self.bottom.load(Relaxed);
+    'L102: let t = self.top.load(Acquire);
+    'L103: let mut buffer = self.buffer.load(Relaxed, epoch::unprotected());
+
+    'L104: let cap = buffer.get_capacity();
+    'L105: if b - t >= cap {
+    'L106:     self.resize(cap * 2);
+    'L107:     buffer = self.buffer.load(Relaxed, epoch::unprotected());
+    'L108: }
+
+    'L109: buffer.write(b % buffer.get_capacity(), t);
+    'L110: self.bottom.store(b + 1, Release);
+}
+
+pub fn pop(&self) -> Option<T> {
+    'L201: let b = self.bottom.load(Relaxed);
+    'L202: self.bottom.store(b - 1, Relaxed);
+    'L203: fence(SeqCst);
+    'L204: let t = self.top.load(Relaxed);
+    'L205: let len = b - t;
+
+    'L206: if len <= 0 {
+    'L207:     self.bottom.store(b, Relaxed);
+    'L208:     return None;
+    'L209: }
+
+    'L210: let buffer = self.buffer.load(Relaxed, epoch::unprotected());
+    'L211: let mut value = Some(buffer.read((b - 1) % buffer.get_capacity()));
+
+    'L212: if len == 1 {
+    'L213:     if self.top.compare_and_swap(t, t + 1, Release, Acquire).is_err() {
+    'L214:         mem::forget(value.take());
+    'L215:     }
+    'L216:     self.bottom.store(b, Relaxed);
+    'L217: } else {
+    'L218:     let cap = buffer.get_capacity();
+    'L219:     if len < cap / 4 {
+    'L220:         self.resize(cap / 2);
+    'L221:     }
+    'L221: }
+
+    'L222: value
+}
+
+fn resize(&self, cap_new) {
+    'L301: let b = self.bottom.load(Relaxed);
+    'L302: let t = self.top.load(Relaxed);
+    'L303: let old = self.buffer.load(Relaxed, epoch::unprotected());
+    'L304: let new = Buffer::new(cap_new);
+
+    'L305: ... // copy data from old to new
+
+    'L306: let guard = epoch::pin();
+    'L307: self.buffer.store(Owned::new(new).into_shared(guard), Release);
+    'L308: guard.defer(move || old.into_owned());
+}
+
+pub fn steal(&self) -> Option<T> {
+    'L401: let mut t = self.top.load(Relaxed);
+
+    'L402: let guard = epoch::pin_fence(); // epoch::pin(), but forces fence(SeqCst)
```

Comment:
What's the difference between `epoch::pin_fence(); // epoch::pin(), but forces fence(SeqCst)` and `epoch::pin`? `pin` already uses `fence(SeqCst)`, but on x86 it uses `lock cmpxchg` instead, is that the problem?

Reply:
`epoch::pin()` is reentrant, and if it is re-entering, it doesn't issue `fence(SeqCst)`.  Here, I needed to issue regardless of whether it is re-entering or not.  Maybe it's not clear from the text.  I'll revise it.